ProjFS.Mac/PrjFSLib/PrjFSLib.cpp

```diff
@@ -162,21 +165,36 @@ static mutex s_fileLocksMutex;
 
 PrjFS_Result PrjFS_RegisterForOfflineIO()
 {
-    s_kernelServiceOfflineWriterConnection = PrjFSService_ConnectToDriver(UserClientType_OfflineIO);
-    if (s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL)
+    mutex_lock lock(s_kernelServiceOfflineClientMutex);
+    
+    if (s_kernelServiceOfflineClientCount == 0)
     {
-        return PrjFS_Result_EDriverNotLoaded;
+        assert(s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL);
+        s_kernelServiceOfflineWriterConnection = PrjFSService_ConnectToDriver(UserClientType_OfflineIO);
+        if (s_kernelServiceOfflineWriterConnection == IO_OBJECT_NULL)
+        {
+            return PrjFS_Result_EDriverNotLoaded;
+        }
     }
+    
+    ++s_kernelServiceOfflineClientCount;
 
     return PrjFS_Result_Success;
 }
 
 PrjFS_Result PrjFS_UnregisterForOfflineIO()
 {
+    mutex_lock lock(s_kernelServiceOfflineClientMutex);
+    
+    assert(s_kernelServiceOfflineClientCount > 0);
```

Comment:
If we have 2 people register for offline, and then 1 unregisters wouldn't we assert?

Reply:
Every successful call to `PrjFS_RegisterForOfflineIO()` increments the count via `++s_kernelServiceOfflineClientCount;`, so if 2 calls to register are made, the count will be 2. I'm not clear on why this would assert. The old code would have asserted on the second unregister.

Reply:
I see so we assert if someone calls `PrjFS_UnregisterForOfflineIO` without a previous pairing  `PrjFS_RegisterForOfflineIO`.
That seems fair.

Reply:
Let's please add a comment above the assert so it's 100% clear

Reply:
I've added comments in both the registration and deregistration functions explaining how each of them deals with the refcounting.